### PR TITLE
Proficiencies

### PR DESCRIPTION
**Pull request by corinnaj**

### CODE REVIEW

File: src/client/components/DicePanel.tsx [hunk 1]
```diff
@@ -1,57 +1,382 @@
...
-  faBroom,
-];
+import { DiceTemplates, GeneratedDiceTemplates } from "./DiceTemplates";
+import { Popover } from "./Popover";
+import { Button } from "./ui/Button";
+import { SmartTextInput } from "./ui/TextInput";
 
 export const DicePanel = React.memo(function DicePanel() {
-  const [active, setActive] = useState(0);
+  const [active, setActive] = useState("Input");
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:42]: This could probably use a better type than string, i.e. `useState<"Input" | "Skills" | "STs" | DiceTemplateCategoryID>`

File: src/client/components/DicePanel.tsx [hunk 2]
```diff
@@ -1,57 +1,382 @@
...
+          category: {
+            id: rrid<RRDiceTemplateCategory>(),
+            icon: freeIcon,
+            categoryName: "",
+            templates: [],
+          },
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:79]:
```suggestion
      syncToServerThrottle: 0,
```
This could probably be 0, since it isn't something that is triggered repeatedly like the `onChange` event of a text input. There is no harm in sending this action to the server immediately.

File: src/client/components/DicePanel.tsx [hunk 3]
```diff
@@ -1,57 +1,382 @@
...
+
+  const d20Part: RRDiceTemplatePartDice = {
+    id: rrid<RRDiceTemplatePart>(),
+    type: "dice",
+    faces: 20,
+    count: 1,
+    negated: false,
+    modified: "none",
+    damage: { type: null },
+  };
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:91]: Can we make this a function, so that we do not reuse the same id for multiple template parts?

File: src/client/components/DicePanel.tsx [hunk 4]
```diff
@@ -1,57 +1,382 @@
...
+  const characters = useServerState((state) => state.characters);
+  const character = characters.entities[myself.mainCharacterId!];
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:93]: Can we move the selection of the main character into the `useServerState` selector, so that we don't unnecessarily subscribe to all character changes? Also, `myself.mainCharacterId!` seems fishy to me - what if I don't have a main character set?

File: src/client/components/DicePanel.tsx [hunk 5]
```diff
@@ -1,57 +1,382 @@
...
+
+  function getSavingThrowTemplates() {
+    const templates = [];
+
+    templates.push(
+      ...characterStatNames.map(
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:99]: The `templates` variable seems unnecessary - you should be able to return the result of the `characterStatNames.map` call directly.

File: src/client/components/DicePanel.tsx [hunk 6]
```diff
@@ -1,57 +1,382 @@
...
+        }
+      )
+    );
+    return templates;
+  }
+
+  function getSkillTemplates() {
+    const templates: RRDiceTemplate[] = [];
+
+    templates.push(
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:138]: `templates` variable is not necessary.

File: src/client/components/DicePanel.tsx [hunk 7]
```diff
@@ -1,57 +1,382 @@
...
+          parts.push({
+            id: rrid<RRDiceTemplatePart>(),
+            type: "linkedProficiency",
+            damage: { type: null },
+            proficiency,
+          });
+
+        return {
+          id: rrid<RRDiceTemplate>(),
+          name: skill,
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:161]: Hmm, would `${skill} Check` be too verbose?

File: src/client/components/DicePanel.tsx [hunk 8]
```diff
@@ -1,57 +1,382 @@
...
+          <>
+            <DiceInterface />
+            <DiceInput />
+          </>
+        );
+      case "Skills":
+        return (
+          <GeneratedDiceTemplates
+            templates={getSkillTemplates()}
+            character={character!}
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:189]: This looks like a bug when a player doesn't have a main character or deletes their main character.

File: src/client/components/DicePanel.tsx [hunk 9]
```diff
@@ -1,57 +1,382 @@
...
+        return (
+          <>
+            <DiceInterface />
+            <DiceInput />
+          </>
+        );
+      case "Skills":
+        return (
+          <GeneratedDiceTemplates
+            templates={getSkillTemplates()}
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:188]: Can we `useMemo` the result of `getSkillTemplates()` so that the generated dice templates are only re-calculated if the character proficiencies change?

I.e. somewhere above:
```ts
const skillTemplates = useMemo(
	() => skillNames.map((skill) => { /* ... */ }),
	[character.proficiencies]
);
```

File: src/client/components/DicePanel.tsx [hunk 10]
```diff
@@ -1,57 +1,382 @@
...
+            character={character!}
+          />
+        );
+      case "STs":
+        return (
+          <GeneratedDiceTemplates
+            templates={getSavingThrowTemplates()}
+            character={character!}
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:196]: Same comment about `useMemo` and the possible bug when not having a main character.

File: src/client/components/DicePanel.tsx [hunk 11]
```diff
@@ -1,57 +1,382 @@
...
+  function deleteCategory() {
+    dispatch({
+      actions: [
+        playerDeleteDiceTemplateCategory({
+          id: myself.id,
+          categoryId: category.id,
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:315]:
```suggestion
      syncToServerThrottle: 0,
```

File: src/client/components/DicePanel.tsx [hunk 12]
```diff
@@ -1,57 +1,382 @@
...
+    newName?: string
+  ) {
+    dispatch({
+      actions: [
+        playerUpdateDiceTemplateCategory({
+          id: myself.id,
+          category: {
+            id: category.id,
+            changes: {
+              icon: newIcon ?? category.icon,
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:331]: This will unnecessarily overwrite the icon with the existing icon if no new icon is set. This might create unnecessary edge cases. We could update the function signature to something like this instead: `categoryId: RRDiceTemplateCategoryID, changes: Partial<Pick<RRDiceTemplateCategory, "icon" | "categoryName">>` and then simply hand over the changes here.

File: src/client/components/DicePanel.tsx [hunk 13]
```diff
@@ -1,57 +1,382 @@
...
+
+function DiceTemplateCategoryEditor({
+  category,
+}: {
+  category: RRDiceTemplateCategory;
+}) {
+  const myself = useMyProps("diceTemplateCategories", "id");
+  const dispatch = useServerDispatch();
+
+  function deleteCategory() {
```
Comment (cmfcmf) [inline src/client/components/DicePanel.tsx:306]: Should we `useConfirm` to ask the user for confirmation before deleting a category?

File: src/client/components/DiceTemplates.tsx [hunk 14]
```diff
@@ -373,6 +392,20 @@ function DicePicker() {
...
+    damage: { type: null },
+    proficiency: 1,
+  };
+
+  const initiativePart: RRDiceTemplatePartLinkedModifier = {
+    id: rrid<RRDiceTemplatePart>(),
+    type: "linkedModifier" as const,
+    damage: { type: null },
+    name: "initiative",
+  };
```
Comment (cmfcmf) [inline src/client/components/DiceTemplates.tsx:407]: Should we make these two functions so that we do not re-use template parts with the same id?

File: src/client/components/DiceTemplates.tsx [hunk 15]
```diff
@@ -854,14 +838,54 @@ function DamageTypeEditor({
...
+  playerUpdateDiceTemplate({
+    id: myId,
+    categoryId,
+    template: {
+      id: diceTemplate.id,
+      changes: {
+        parts: diceTemplate.parts.filter((p) => p.id !== partId),
+      },
+    },
+  });
```
Comment (cmfcmf) [inline src/client/components/DiceTemplates.tsx:856]: Does this work with nested templates?

File: src/client/components/DiceTemplates.tsx [hunk 16]
```diff
@@ -854,14 +838,54 @@ function DamageTypeEditor({
...
+    categoryId,
+    template: {
+      id: diceTemplate.id,
+      changes: {
+        parts: diceTemplate.parts.map((p) =>
+          p.id === partId ? { ...p, changes } : p
+        ),
+      },
+    },
+  });
```
Comment (cmfcmf) [inline src/client/components/DiceTemplates.tsx:876]: Does this work with nested templates?

File: src/client/components/DiceTemplates.tsx [hunk 17]
```diff
@@ -954,60 +978,112 @@ function ModifierNumberEditor({
...
+        }
+      />
+    </label>
   );
+}
 
-  if (!template) {
-    return null;
-  }
+function TemplateNoteEditor({
```
Comment (cmfcmf) [inline src/client/components/DiceTemplates.tsx:1020]: This should probably be renamed to something like `TemplateSettingsEditor`.

File: src/client/components/DiceTemplates.tsx [hunk 18]
```diff
@@ -954,60 +978,112 @@ function ModifierNumberEditor({
...
+            key={category.id}
+            className="radio-label"
+            title={category.categoryName}
+          >
+            <input
+              type="radio"
+              name="icon"
+              value={category.id}
+              //TODO how was this working?
+              checked={categoryId === category.id}
```
Comment (cmfcmf) [inline src/client/components/DiceTemplates.tsx:1045]: Hmm?

File: src/client/components/DiceTemplates.tsx [hunk 19]
```diff
@@ -954,60 +978,112 @@ function ModifierNumberEditor({
...
+                    //TODO does this work?
+                    playerRemoveDiceTemplate({
+                      id: myself.id,
+                      categoryId: category.id,
+                      templateId: template.id,
+                    }),
+                    playerAddDiceTemplate({
+                      id: myself.id,
+                      categoryId: category.id,
+                      template: template,
```
Comment (cmfcmf) [inline src/client/components/DiceTemplates.tsx:1058]: Good question :D

**Review by cmfcmf — COMMENTED** [review]
Looks really good and like a lot of work! 🎉 
I added a few inline comments. I wasn't able to test the code in a browser yet because the migration is failing for me, likely due to nested dice templates.

File: src/client/myself.tsx [hunk 20]
```diff
@@ -76,16 +77,26 @@ export function MyselfProvider({ children }: { children: React.ReactNode }) {
   );
```
[the remark this answers is not in the document]
Comment (corinnaj) [reply src/client/myself.tsx:77]: The useMyId() function is currently also used inside useMyProps(). If I remove it, do I use useMyself() there, or directly get it from the server state as in the MyselfProvider? Does either impact the performance of the function?

File: src/server/migrations/19_add_proficiencies.ts [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+          diceTemplateCategories: {
+            categoryName: string;
+            icon: any;
+            templates: [];
+            id: string;
+          }[];
+        }
+      >
+    ).forEach((player) => {
+      //TODO Corinna
```
[the remark this answers is not in the document]
Comment (corinnaj) [reply src/server/migrations/19_add_proficiencies.ts:65]: Ups, yeah, was trying to create a note that's easy to find, there are more TODOs in our code than I expected :D

File: src/client/components/DicePanel.tsx (hunk 7 above)
Comment (corinnaj) [reply src/client/components/DicePanel.tsx:161]: I think at least in the Notification it has a nicer ring without it: "xx rolled for Athletics..."

File: src/client/myself.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (corinnaj) [reply src/client/myself.tsx:77]: Or is `useContext(MyselfContext)` as it was previously in the `useMyself()` best here? With the addition of throwing an error when `player` is `null`?

File: src/client/components/DicePanel.tsx (hunk 7 above)
Comment (cmfcmf) [reply src/client/components/DicePanel.tsx]: Fair point!

File: src/client/state.tsx [hunk 22]
```diff
@@ -543,14 +551,23 @@ export function useServerStateRef<T>(
...
+      ) {
         selectedStateRef.current = newSelectedState;
         onChangeRef.current?.(newSelectedState);
       }
     };
     subscribe(subscriber);
     return () => unsubscribe(subscriber);
-  }, [onChangeRef, selectedStateRef, selectorRef, subscribe, unsubscribe]);
+  }, [
+    equalityFn,
```
[the remark this answers is not in the document]
Comment (cmfcmf) [reply src/client/state.tsx:564]: After re-reading my comment, I realize that it might not be super clear. Let me try again:
The first effect is supposed to subscribe to every server state change. On every change, it applies the selector and compares the selected part of state with the previously selected part of state using equalityFn. 
We basically never want to unsubscribe and resubscribe in this effect, which is why all dependencies (except for subscribe and unsubscribe themselves) should be refs.

However, what if equalityFn or the selector changes? Let's take this example:

```ts
function MyComponent({id}: {id: RRMapID}) {
  const map = useServerState(state => state.maps.entities[id]);

  // ... 
}
```

In this example, the selector is dependent on a prop. If we only had the first effect, the `map` would not update when the `id` passed to this component changes. That is what the second effect further down in `useServerStateRef` is for. It should run whenever `selector` or `equalityFn` changes (also see my other comment: https://github.com/cmfcmf/roc-and-roll/pull/137#discussion_r748877400) and checks if the selected part of the state is still equal to the previously selected part. If not, it executes the onChange callback. This is why we want the second effect to re-run whenever `selector` or `equalityFn` changes.

File: src/client/myself.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (cmfcmf) [reply src/client/myself.tsx:42]: > The `useMyId()` function is currently also used inside `useMyProps()`.

Ah, I missed that.

> Or is `useContext(MyselfContext)` as it was previously in the `useMyself()` best here? With the addition of throwing an error when `player` is `null`?

I think that sounds good, if we also modify `MyselfContext` to no longer contain the entire player object but just the player id or `null`.

File: src/client/components/DiceTemplates.tsx [hunk 23]
```diff
@@ -1151,13 +1246,27 @@ const DiceTemplatePart = React.forwardRef<
         </div>
       );
       break;
+    case "linkedProficiency":
+      content = (
+        <div className="dice-option" style={styleFor(part)}>
+          <div className="dice-option-linked-modifier">
+            {
+              //TODO this shows incorrect value in editor, others show no value
```
[the remark this answers is not in the document]
Comment (corinnaj) [reply src/client/components/DiceTemplates.tsx:1227]: This was an issue, when the corresponding value wasn't set in the selected character, should be fixed now

File: src/client/components/DicePanel.tsx (hunk 13 above)
Comment (corinnaj) [reply src/client/components/DicePanel.tsx:294]: I am a bit hesitant to use pop ups, but I guess in this case it might make sense. Before people start accidentally deleting all their templates.

File: src/client/components/DiceTemplates.tsx (hunk 14 above)
Comment (corinnaj) [reply src/client/components/DiceTemplates.tsx]: Okay, I made them into functions now, but had to assign them as we need a reference later. Is that fine?

File: src/client/components/DicePanel.tsx (hunk 13 above)
Comment (cmfcmf) [reply src/client/components/DicePanel.tsx:294]: I am also always in favor of a nicer UI that doesn't rely on popups 👍 To me, popups are mostly a placeholder for 'we haven't made the UX great yet'. 
It's just that I would be super sad if I accidentally deleted my dice templates^^

Comment (corinnaj) [reply src/client/components/DicePanel.tsx:294]: So for all of our UX, yes? :smile:

File: src/client/components/DiceTemplates.tsx (hunk 14 above)
Comment (cmfcmf) [reply src/client/components/DiceTemplates.tsx]: Oh, I'm sorry, After looking at the code again, it seems like it was just fine the way you had it initially 🙈 since we overwrite these ids before adding parts to a proper template:
https://github.com/cmfcmf/roc-and-roll/blob/ced4c7ca91ea7659ef0aad0cc4180512c181a436/src/client/components/DiceTemplates.tsx#L99
https://github.com/cmfcmf/roc-and-roll/blob/ced4c7ca91ea7659ef0aad0cc4180512c181a436/src/client/components/DiceTemplates.tsx#L568-L593

I have some thoughts on how to make this less likely to break in the future, but that's unrelated to this PR. I think we can just leave it as it is for now.

File: src/client/components/DicePanel.tsx (hunk 13 above)
Comment (cmfcmf) [reply src/client/components/DicePanel.tsx:294]: Psssssst... don't tell anyone! :D

File: src/client/myself.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (corinnaj) [reply src/client/myself.tsx:42]: Okay, I think I changed the `MyselfContext` so that is only contains the `playerId`. However, I seem to have broken prompts in the meantime? Could that be related to this somehow? (Prompt never show up, not for log in neither for creating dice templates from the roll interface)

[the remark this answers is not in the document]
Comment (cmfcmf) [reply src/client/myself.tsx:42]: Ummm, I don't see how these could be related, weird. But `myself.tsx` looks good to me now 🎉

[the remark this answers is not in the document]
Comment (corinnaj) [reply src/client/myself.tsx:42]: Unfortunate :see_no_evil:

File: src/client/components/DicePanel.tsx [hunk 24]
```diff
@@ -1,57 +1,382 @@
...
+            id: category.id,
+            changes: {
+              icon: newIcon ?? category.icon,
+              categoryName: newName ?? category.categoryName,
+            },
+          },
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
```
Comment (corinnaj) [inline src/client/components/DicePanel.tsx:326]: Should this then also be changed to 0?

Comment (cmfcmf) [reply src/client/components/DicePanel.tsx:326]: I don't think so, because it is triggered by the text input that changes the categoryName. We wouldn't want to send an update to the server on every character you type into the text input.

Comment (corinnaj) [reply src/client/components/DicePanel.tsx:326]: Ah, fair enough

File: src/shared/features/players.ts [hunk 25]
```diff
@@ -32,6 +38,58 @@ export const playersReducer = createReducer(
...
+        const index = player?.diceTemplateCategories.findIndex(
+          (c) => c.id === action.payload.categoryId
+        );
+        if (index !== undefined && index >= 0) {
+          player?.diceTemplateCategories.splice(index, 1);
+        }
+      })
+
+      .addCase(playerRemoveDiceTemplate, (state, action) => {
+        // TODO go recursive
```
[the remark this answers is not in the document]
Comment (corinnaj) [reply src/shared/features/players.ts]: Hopefully now

File: src/client/components/DiceTemplates.tsx (hunk 19 above)
Comment (corinnaj) [reply src/client/components/DiceTemplates.tsx]: It should now. Might be a more elegant way of doing this though...

File: src/client/components/DiceTemplates.tsx (hunk 14 above)
Comment (corinnaj) [reply src/client/components/DiceTemplates.tsx:388]: Alright, will undo it for now then

File: src/client/state.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (tom95) [reply src/client/state.tsx]: Changes are now in. Would be great if you could take another look! The levels of indirection certainly contributed to some confusion :D

File: src/client/state.tsx [hunk 26]
```diff
@@ -571,11 +570,11 @@ export function useServerStateRef<T>(
 
   useEffect(() => {
     const newSelectedState = selector(stateRef.current);
-    if (!Object.is(newSelectedState, selectedStateRef.current)) {
+    if (!equalityFn(newSelectedState, selectedStateRef.current)) {
```
Comment (cmfcmf) [inline src/client/state.tsx:573]: We need to swap the function parameters here :)

File: src/client/state.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (cmfcmf) [reply src/client/state.tsx]: Looks good now except for a minor issue (see other comment). Great work!

**Comment by cmfcmf** [comment]
I noticed two more issues:

- the right click menu of nested dice templates shows buttons to select the category, which don't do anything -> those buttons shouldn't be there for nested templates
- deleting a nested dice template from the right click menu deletes the entire template

File: src/shared/validation.ts [hunk 27]
```diff
@@ -164,6 +162,128 @@ export const isStateVersion = t.applyCascade(t.isNumber(), [
...
+
+export type RRDiceTemplate = t.InferType<typeof isDiceTemplate>;
+
+assert<IsExact<RRDiceTemplate, _RRDiceTemplate>>(true);
+
+const isDiceTemplateCategory = t.isObject({
+  id: isRRID<RRDiceTemplateCategoryID>(),
+  icon: t.isEnum(categoryIcons),
+  categoryName: t.isString(),
+  templates: t.isArray(isDiceTemplate),
```
Comment (cmfcmf) [inline src/shared/validation.ts:283]: After migrating my state, I have some dice templates with the same id now. This might totally be because I my state was in a weird state that the migration does not account for and it might not happen with our production state. Could we adjust the validation here to check for duplicate ids and test it on our production state?

```ts
  templates: t.applyCascade(t.isArray(isDiceTemplate), [
    t.hasUniqueItems({ map: (diceTemplate) => diceTemplate.id }),
  ]),
```

**Comment by corinnaj** [comment]
Should all be fixed now :)

**Review by cmfcmf — APPROVED** [review]
Looks great now, can't wait to use this when we play! 🎉
